### PR TITLE
Validate ASCII host values

### LINKED ISSUE #880: basic-auth user and password are converted to lowercase when passed as part of "host" to URL.build
### Describe the bug

user and password casing is preserved when using the constructor:

    >>> yarl.URL('httPS://usER:passWORD@hostNAME/paTH')
    URL('https://usER:passWORD@hostname/paTH')
    >>> yarl.URL('httPS://usER:passWORD@hostNAME/paTH').user
    'usER'

user and password casing is lost when using `URL.build`:

    >>> yarl.URL.build(scheme='httPS', host='usER:passWORD@hostNAME', path='/paTH')
    URL('httPS://user:password@hostname/paTH')
    >>> yarl.URL.build(scheme='httPS', host='usER:passWORD@hostNAME', path='/paTH').user
    'user'

Note that the user & password are extracted from the string passed as `host`.

passing user and password explicitly also preserves the casing:

    >>> yarl.URL.build(scheme='httPS', user='usER', password='passWORD', host='hostNAME', path='/paTH')
    URL('httPS://usER:passWORD@hostname/paTH')
    >>> yarl.URL.build(scheme='httPS', user='usER', password='passWORD', host='hostNAME', path='/paTH').user
    'usER'

I'd expect either user and password to preserve casing when passed into `host` (preferably) or them not being extracted into the `user` and `password` attributes and instead remain in the lowercase and encoded `host` attribute.

maybe caused by https://github.com/aio-libs/yarl/issues/386

converting the host to lowercase should be done after extracting username / password.

### To Reproduce

    >>> yarl.URL.build(scheme='httPS', host='usER:passWORD@hostNAME', path='/paTH')
    URL('httPS://user:password@hostname/paTH')
    >>> yarl.URL.build(scheme='httPS', host='usER:passWORD@hostNAME', path='/paTH').user
    'user'

### Expected behavior

    >>> yarl.URL.build(scheme='httPS', host='usER:passWORD@hostNAME', path='/paTH')
    URL('httPS://user:password@hostname/paTH')
    >>> yarl.URL.build(scheme='httPS', host='usER:passWORD@hostNAME', path='/paTH').user
    'usER'

### Logs/tracebacks

```python-traceback
N/A
```

### Python Version

```console
$ python --version
Python 3.10.10
```

### multidict Version

```console
$ python -m pip show multidict
Name: multidict
Version: 6.0.4
Summary: multidict implementation
Home-page: https://github.com/aio-libs/multidict
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /usr/lib/python3.10/site-packages
Requires: 
Required-by: aiohttp, httpie, yarl
```

### yarl Version

```console
$ python -m pip show yarl
Name: yarl
Version: 1.8.2
Summary: Yet another URL library
Home-page: https://github.com/aio-libs/yarl/
Author: Andrew Svetlov
Author-email: andrew.svetlov@gmail.com
License: Apache 2
Location: /home/stefan/.local/lib/python3.10/site-packages
Requires: idna, multidict
Required-by: aiohttp
```

### OS

Arch Linux

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow the aio-libs Code of Conduct

### PR DESCRIPTION
**Pull request by mjpieters**
Host values must match the unreserved, sub-delims or pct-encoded grammar
rules. The IDNA encoder already checks for this, but for ASCII values we
skip IDNA encoding.

If the invalid character is `@`, or `:` followed by `@` further in the
host value, tack on advice about using `authority` instead of `host`.

Fixes #880

Screenshot 2024-09-07 at 5 19 09 PM

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` (e.g. `588.bugfix`)
  * if you don't have an `issue_id` change it to the pr id after creating the PR
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: `Fix issue with non-ascii contents in doctest text files.`

### CODE REVIEW

File: CHANGES/8080.bugfix.rst
[the remark this answers is not in the document]
Comment (webknjaz) [reply CHANGES/8080.bugfix.rst:1]: Also, we can optionally symlink it to `954.bugfix.rst` so both PR and issue links are included in the changelog.

File: CHANGES/8080.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Started rejecting ASCII hostnames with invalid characters. For host strings that
+look like authority strings, the exception message includes advice on what to do
+instead.
```
Comment (webknjaz) [inline CHANGES/8080.bugfix.rst:3]: I've added bylines to all the present changelog fragments. Let's add one here too, for consistency:
```suggestion
instead -- by :user:`mjpieters`.
```

The preview will be here: https://yarl--954.org.readthedocs.build/en/954/changes.html

File: CHANGES/8080.bugfix.rst
[the remark this answers is not in the document]
Comment (mjpieters) [reply CHANGES/8080.bugfix.rst:1]: oops, will fix!

File: CHANGES/8080.bugfix.rst (hunk 1 above)
Comment (mjpieters) [reply CHANGES/8080.bugfix.rst:3]: Sure thing!

**Comment by webknjaz** [comment]
@mjpieters do you want me to reject the release build, wait and include this PR?

**Comment by webknjaz** [comment]
> ```console
> [towncrier-fragments]:13: : Spell check: hostnames: Started rejecting ASCII hostnames with invalid characters. For host strings that.
> ```

@mjpieters this can be fixed by adding "hostnames" to `docs/spelling_wordlist.txt`. Though, I hope to integrate `codespell` later that would replace `sphinxcontrib-spelling` and seems to have less false positives.

**Comment by mjpieters** [comment]
While implementing my PR I noticed that if you used a non-ASCII hostname with `@` or any other invalid character (so not unreserved, percent encoded or sub-delims), then `idna.encode()` would raise an exception. But when I tried to reproduce this behaviour, just now, I noticed **we swallow that exception**, because it is a `UnicodeError` subclass 🤦 .

I'm not sure why the encoder uses `except UnicodeError` with a fallback to `host.encode('idna').decode('ascii')`, because that completely defeats the valid codepoint check.

I can hoist the validation to apply to non-ASCII hostnames too but that means I need to explicitly test for all disallowed ASCII codepoint (instead of the negative character classes I use now).

I'll go search through the commit history to see if I cannot find out more here.

**Comment by mjpieters** [comment]
> @mjpieters this can be fixed by adding "hostnames" to `docs/spelling_wordlist.txt`. Though, I hope to integrate `codespell` later that would replace `sphinxcontrib-spelling` and seems to have less false positives.

:-D I already have that change staged locally :-P

**Comment by webknjaz** [comment]
@mjpieters so do you think it's worth putting into a separate release? The automation is there so it wouldn't take as long to make new releases from now on...

**Comment by mjpieters** [comment]
> Not sure why the changelog bot is failing to notice the towncrier fragment... Not sure why but it's safe to ignore for now.

It is looking for the files in the wrong location, `news/` instead of `CHANGES/`.

**Comment by mjpieters** [comment]
> @mjpieters do you want me to reject the release build, wait and include this PR?

I think it's fine to put this in a next release. I don't see this as urgent.

**Comment by webknjaz** [comment]
> > @mjpieters do you want me to reject the release build, wait and include this PR?
> 
> I think it's fine to put this in a next release. I don't see this as urgent.

Okay, will proceed with releasing, then. The build is half-way ready: https://github.com/aio-libs/yarl/actions/runs/6931722870.

File: CHANGES/880.bugfix.rst
Comment (webknjaz) [inline CHANGES/880.bugfix.rst:1]: @mjpieters did you opt for a copy instead of a symlink? These are usually harder to maintain when the exact same edits are needed to both files... Git handles symlinks well (although, I know that it turns them into regular files on Windows when checking out).

**Comment by mjpieters** [comment]
> Okay, will proceed with releasing, then. The build is half-way ready: https://github.com/aio-libs/yarl/actions/runs/6931722870.

What a glorious deployment pipeline we have! 😁

File: CHANGES/880.bugfix.rst
Comment (mjpieters) [reply CHANGES/880.bugfix.rst:1]: No, I meant to use a symlink. I guess I mucked up and made a hardlink.

**Comment by webknjaz** [comment]
> > Not sure why the changelog bot is failing to notice the towncrier fragment... Not sure why but it's safe to ignore for now.
> 
> It is looking for the files in the wrong location, `news/` instead of `CHANGES/`.

Oh, that makes sense — this is because I moved the config to a dedicated file and haven't updates the bot to account for that. Thanks for pointing it out!

**Comment by webknjaz** [comment]
> > Okay, will proceed with releasing, then. The build is half-way ready: https://github.com/aio-libs/yarl/actions/runs/6931722870.
> 
> What a glorious deployment pipeline we have! 😁

Hopefully, I'll make it scalable and reuse for other projects. I have similar things in many other repos but they all have tiny differences which makes them hard to reuse as is.
I made the CI run tests from the exact sdists we publish and install the wheels instead of rebuilding from Git source in each job. Though, I broke codecov reporting along the way so ignore it for now, I'll fix it later, hopefully.

**Comment by webknjaz** [comment]
> Okay, will proceed with releasing, then. The build is half-way ready: https://github.com/aio-libs/yarl/actions/runs/6931722870.

It took a few reattempts to get the last bit right... But now we have v1.9.3:
- https://pypi.org/project/yarl/1.9.3/
- https://github.com/aio-libs/yarl/discussions/957
- https://github.com/aio-libs/yarl/releases/tag/v1.9.3

**Comment by webknjaz** [comment]
> Though, I broke codecov reporting along the way so ignore it for now, I'll fix it later, hopefully.

Fixed it. The coverage metrics are back up!

**Comment by webknjaz** [comment]
> I'm not sure why the encoder uses `except UnicodeError` with a fallback to `host.encode('idna').decode('ascii')`, because that completely defeats the valid codepoint check.

I checked the coverage reported before I broke it temporarily, and turns out that the line you're talking about is _the only one uncovered_ in the entire project: https://app.codecov.io/gh/aio-libs/yarl/commit/65333fab5f5b2bcce9fbb6aa810e9618fa10c6f0/blob/yarl/_url.py#L1176.

So I guess this might also need some extra extensive testing...

File: yarl/_url.py [hunk 2]
```diff
@@ -778,11 +795,27 @@ def _encode_host(cls, host, human=False):
...
-            # IDNA encoding is slow,
-            # skip it for ASCII-only strings
+            # skip it for humanized and ASCII-only strings
             # Don't move the check into _idna_encode() helper
             # to reduce the cache size
-            if human or host.isascii():
+            if human:
+                return host
+            if host.isascii():
+                # Check for invalid characters explicitly; _idna_encode() does this
```
Comment (bdraco) [inline yarl/_url.py:804]: Should this get an LRU like `_idna_encode`?

**Comment by webknjaz** [comment]
> the line you're talking about is _the only one uncovered_ in the entire project: [app.codecov.io/gh/aio-libs/yarl/commit/65333fab5f5b2bcce9fbb6aa810e9618fa10c6f0/blob/yarl/_url.py#L1176](https://app.codecov.io/gh/aio-libs/yarl/commit/65333fab5f5b2bcce9fbb6aa810e9618fa10c6f0/blob/yarl/_url.py#L1176).

This is no longer true since I implemented collecting Cython and MyPy coverage...

File: yarl/_url.py (hunk 2 above)
Comment (bdraco) [reply yarl/_url.py]: Maybe... its 3x more expensive than anything else in this function

File: tests/test_url_build.py [hunk 3]
```diff
@@ -21,7 +21,7 @@ def test_url_build_ipv6():
 
 
 def test_url_build_ipv6_brackets():
-    u = URL.build(scheme="http", host="[::1]")
+    u = URL.build(scheme="http", host="[::1]", encoded=True)
```
Comment (bdraco) [inline tests/test_url_build.py:24]: This test was incorrect. Its already encoded

File: yarl/_url.py [hunk 4]
```diff
@@ -943,8 +961,12 @@ def _encode_host(cls, host: str, human: bool = False) -> str:
         # skip it for ASCII-only strings
         # Don't move the check into _idna_encode() helper
         # to reduce the cache size
-        if human or host.isascii():
+        if human:
             return host
+        if host.isascii():
+            # Check for invalid characters explicitly; _idna_encode() does this
+            # for non-ascii host names.
```
Comment (bdraco) [inline yarl/_url.py:968]:
```suggestion
```

File: yarl/_url.py [hunk 5]
```diff
@@ -943,8 +961,12 @@ def _encode_host(cls, host: str, human: bool = False) -> str:
...
+            _host_validate(host)
```
Comment (bdraco) [inline yarl/_url.py:969]:
```suggestion
            _host_validate(host)
            return host
```

File: yarl/_url.py [hunk 6]
```diff
@@ -255,7 +273,7 @@ def __new__(
                 username, password, host, port = cls._split_netloc(val[1])
                 if host is None:
                     raise ValueError("Invalid URL: host is required for absolute urls")
-                host = cls._encode_host(host)
+                host = cls._encode_host(host, validate_host=False)
```
Comment (bdraco) [inline yarl/_url.py:276]: We split / parsed the url so it never going to have the same concerns as #880 so I turned it off here.  There is different validation in urllib which raises different errors so I added a test for that. We may need to do a followup to smooth out this rough edge, but since the case in #880 is addressed, I figure this is good enough for a first pass.

File: tests/test_url_build.py (hunk 3 above)
Comment (bdraco) [inline tests/test_url_build.py:24]: I'm a bit worried about having to change this test so I'm going to leave it for someone else to take a look before merging, otherwise I think its ready to go now.

**Review by bdraco — APPROVED** [review]

File: tests/test_url_build.py [hunk 7]
```diff
@@ -21,7 +21,7 @@ def test_url_build_ipv6():
...
     assert str(u) == "http://::1"
```
Comment (bdraco) [inline tests/test_url_build.py:25]:
```suggestion
    assert str(u) == "http://[::1]"
```

File: tests/test_url_build.py (hunk 3 above)
Comment (bdraco) [reply tests/test_url_build.py:24]: Once I fix https://github.com/aio-libs/yarl/pull/1158 if this still passes after re-basing, I'm now confident enough to merge this

**Review by bdraco — APPROVED** [review]

**Comment by bdraco** [comment]
Ironically we were even doing it wrong in `aiohttp` https://github.com/aio-libs/aiohttp/issues/9307
